### PR TITLE
Update React-Redux client app to use TypeScript instead of JavaScript

### PR DESCRIPTION
**Pull request by martinpe36**
We used npx create-react-app --typescript and then updated the current React-Redux template to use the TypeScript files instead of the JavaScript files

### CODE REVIEW

**Comment by Eilon** [comment]
@SteveSandersonMS / @ryanbrandenburg / @mkArtakMSFT  - thoughts on this? It's a pretty huge change.

**Comment by SteveSandersonMS** [comment]
For context on this PR, see https://github.com/aspnet/Templating/pull/816. It looks like @ryanbrandenburg and @martinpe36 were in the middle of a code review conversation there.

**Comment by ryanbrandenburg** [comment]
@martinpe36 it looks like your PR is broken by an external change, pull the latest into your branch and see if that fixes it.

**Comment by PFAR** [comment]
@martinpe36 
If you update to connected-react-router 6.1.0, configureStore.ts becomes:
```typescript
import { applyMiddleware, combineReducers, compose, createStore, Reducer } from 'redux';
import thunk from 'redux-thunk';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { History } from 'history';
import { ApplicationState, reducers } from './';

export default function configureStore(history: History, initialState?: ApplicationState) {
    const middleware = [
        thunk,
        routerMiddleware(history)
    ];

    const rootReducer = combineReducers({
        ...reducers,
        router: connectRouter(history)
    });

    const enhancers = [];
    const windowIfDefined = typeof window === 'undefined' ? null : window as any;
    if (windowIfDefined && windowIfDefined.__REDUX_DEVTOOLS_EXTENSION__) {
        enhancers.push(windowIfDefined.__REDUX_DEVTOOLS_EXTENSION__());
    }

    return createStore(
        rootReducer,
        initialState,
        compose(applyMiddleware(...middleware), ...enhancers)
    );
}
```
Also noticed you referenced react-router-redux in the original configureStore.ts although it's not in package.json (also good call not to include it as it's depricated).

This is the package.json after upgrading:
```json
{
  "name": "WebApplication1",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "bootstrap": "4.2.1",
    "connected-react-router": "6.1.0",
    "jquery": "3.3.1",
    "merge": "1.2.1",
    "history": "4.7.2",
    "react": "16.7.0",
    "react-dom": "16.7.0",
    "react-redux": "6.0.0",
    "react-router": "4.3.1",
    "react-router-dom": "4.3.1",
    "react-scripts": "2.1.2",
    "reactstrap": "6.5.0",
    "redux": "4.0.1",
    "redux-thunk": "2.3.0",
    "typescript": "3.2.2"
  },
  "devDependencies": {
    "@types/history": "4.7.2",
    "@types/jest": "23.3.11",
    "@types/node": "10.12.18",
    "@types/react": "16.7.18",
    "@types/react-dom": "16.0.11",
    "@types/react-redux": "6.0.11",
    "@types/react-router": "4.4.3",
    "@types/react-router-dom": "4.3.1",
    "@types/reactstrap": "6.4.3",
    "cross-env": "5.2.0",
    "eslint": "5.6.0",
    "eslint-config-react-app": "3.0.6",
    "eslint-plugin-flowtype": "2.50.3",
    "eslint-plugin-import": "2.14.0",
    "eslint-plugin-jsx-a11y": "6.1.2",
    "eslint-plugin-react": "7.11.1",
    "popper.js": "1.14.6",
    "typescript-eslint-parser": "21.0.2"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "cross-env CI=true react-scripts test --env=jsdom",
    "eject": "react-scripts eject",
    "lint": "eslint ./src/**/*.ts ./src/**/*.tsx"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": [
    ">0.2%",
    "not dead",
    "not ie <= 11",
    "not op_mini all"
  ]
}
```
Also had to modify app.test.tsx to suppress TypeScript complaints:
```typescript
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import App from './App';

it('renders without crashing', () => {
    const storeFake = (state: any) => ({
        default: () => { },
        subscribe: () => { },
        dispatch: () => { },
        getState: () => ({ ...state })
    }) as any;
    const store = storeFake({});

    ReactDOM.render(
        <Provider store={store}>
            <MemoryRouter>
                <App />
            </MemoryRouter>
        </Provider>, document.createElement('div'));
});
```

Looking forward to seeing TypeScript back in the default react(/redux) templates!!!

**Comment by mkArtakMSFT** [comment]
@ryanbrandenburg, we should consider this PR for preview3 release. In the meantime I appreciate your and @SteveSandersonMS's review here.

**Comment by mkArtakMSFT** [comment]
@javiercn, would be great if you could find time to review this too, as you've recently did similar work in some other area.

**Comment by PFAR** [comment]
Perhaps also add some Typescript best practices similar to the v. 2.0 templates, like the error when forgetting to implement all cases in switch.
`
default:
            // The following line guarantees that every action in the KnownAction union has been covered by a case above
            const exhaustiveCheck: never = action;
            if (typeof exhaustiveCheck != 'undefined') break;

`

File: src/Templating/src/Microsoft.DotNet.Web.Spa.ProjectTemplates/content/ReactRedux-CSharp/ClientApp/package.json [hunk 1]
```diff
@@ -3,40 +3,55 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
-    "bootstrap": "^4.1.3",
+    "bootstrap": "4.2.1",
+    "connected-react-router": "6.2.1",
```
Comment (javiercn) [inline src/Templating/src/Microsoft.DotNet.Web.Spa.ProjectTemplates/content/ReactRedux-CSharp/ClientApp/package.json]: This is a new package @Eilon

Comment (Eilon) [reply src/Templating/src/Microsoft.DotNet.Web.Spa.ProjectTemplates/content/ReactRedux-CSharp/ClientApp/package.json]: Go ahead. We'll use Component Governance to register it.

**Comment by martinpe36** [comment]
I merged from the upstream master branch, but somehow messed up the PR.  I'll fix this and update the PR again.

**Comment by martinpe36** [comment]
I managed to fix up the pull request with the correct changes.  Can you take another look to ensure it's OK?  I haven't updated the package versions from the comment above.  Are these still needed?

**Comment by ryanbrandenburg** [comment]
Your update looks good, I'll merge it as long as tests pass.

@javiercn's comment wasn't about the specific versions you were using, it's an internal dependency  tracking thing. No need to change versions.

**Comment by ryanbrandenburg** [comment]
Sorry about the delay, between weather and vacation I'm a bit behind. Merging!

**Pull request merged**